`ultralytics 8.3.29` Sony IMX500 export

Laughing-q opened the pull request:
@ambitious-octopus here!
working with [this script](https://github.com/ambitious-octopus/model_optimization/blob/main/ultralytics-exports/vanilla-yolo.py).

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This update introduces support for exporting Ultralytics YOLOv8 models to the Sony IMX500 format, enhancing integration capabilities for AI deployment on Raspberry Pi AI Cameras.

### 📊 Key Changes
- Added Sony IMX500 format to the export options within the Ultralytics YOLO framework.
- Updated `.gitignore` to include `*_imx_model/` for ignoring generated IMX model files.
- Documentation enhancements introducing a guide for exporting to Sony IMX500.
- Implementation of a new `FXModel` class for better torch.fx compatibility.
- Added test for IMX export functionality, with specific conditions for running it.
- Updated export and backend logic to support IMX format, including customized model conversion conditions.

### 🎯 Purpose & Impact
- **Purpose:** To expand the range of hardware deployment options, allowing users to optimize and deploy Ultralytics models on IMX500-enabled Raspberry Pi AI Cameras.
- **Impact:** Improves flexibility and utilization of YOLO models for low-power, high-performance edge applications, particularly in robotics and IoT projects featuring AI-enabled cameras. This broadens the potential use cases and improves hardware integration, contributing to more efficient machine learning workflows in edge devices.

Laughing-q committed:
update

Laughing-q committed:
Update block.py

Laughing-q committed:
Update tasks.py

Laughing-q committed:
update

Laughing-q committed:
update

Laughing-q committed:
Update tasks.py

Laughing-q committed:
Update head.py

Laughing-q committed:
Update head.py

Laughing-q committed:
Update head.py

Laughing-q committed:
Update head.py

Laughing-q committed:
Update head.py

Laughing-q committed:
Update head.py

Laughing-q committed:
Update tal.py

Laughing-q committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q commented on ultralytics/nn/tasks.py:
```diff
@@ -87,7 +87,7 @@
 class BaseModel(nn.Module):
     """The BaseModel class serves as a base class for all the models in the Ultralytics YOLO family."""
 
-    def forward(self, x, *args, **kwargs):
```
Removing `*args, **kwargs` arguments to make `torch.fx` work properly. If I don't miss anything It'll be completely compatible with our other pipelines if we can figure a way to add `*args, **kwargs` back without breaking `torch.fx` workflow.

Laughing-q replied:
resolved by adding a `decorator`, but honestly I don't like it😂

Laughing-q replied:
@glenn-jocher ah yeah honestly I don't like it either...but we have to add it here to model.forward process to make torch.fx work. And actually it'll only affect the behavior when using torch.fx by checking if it's `is_fx_tracing()`.

Laughing-q replied:
@glenn-jocher I remember I tried several methods that trying to avoid adding it here i.e. add another forward method without any `*args` and `**kwargs` and try replacing the original one by it right before exporting, like we did for the c2f module ,but those just don't work.

Laughing-q replied:
or tried not getting rid of `*args` and `**kwargs` at all by explicitly using all the arguments `forward` might need so it won't impact other formats. But does not work with torch.fx.

Laughing-q committed:
Update tasks.py

Laughing-q committed:
Update tal.py

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference Section by https://ultralytics.com/actions

Laughing-q committed:
Update tasks.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

ambitious-octopus commented on ultralytics/nn/modules/head.py:
```diff
@@ -89,7 +91,14 @@ def _inference(self, x):
+                for x in make_anchors(getattr(self, "feats_size", torch.Tensor([80, 40, 20])), self.stride, 0.5)
+            )
+            self.anchors, self.strides = self.anchors.to(x[0].device), self.strides.to(x[0].device)
+            self.strides /= 640  # NOTE: this part could be removed in the future.
```
@Laughing-q  also this can be removed!

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

Laughing-q committed:
Update head.py

ambitious-octopus committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

Laughing-q committed:
Update head.py

ambitious-octopus committed:
added mct quantization

ambitious-octopus committed:
added sony post-processing

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
Update head.py

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus commented:
Hey @Laughing-q I've integrated MCT quantization and the exporter. However, I haven't included the quantized model validation, as I believe it may not be necessary in this context. What are your thoughts?

Test with:
```python
from ultralytics import YOLO
model = YOLO("yolov8n.pt")
model.export(format="mct", batch=2)
```

ambitious-octopus committed:
added gptq exporter

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

ambitious-octopus committed:
added gptq export

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
added mct AutoBackend

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus reviewed (commented):
Hey @Laughing-q Great news! I was able to successfully load the exported ONNX models using MCT. You can test it with the following code:

```python
from ultralytics import YOLO
model = YOLO("yolov8n.pt")
model.export(format="mct", device=0)
model = YOLO("yolov8n_mct_model.onnx")
res = model.predict("https://ultralytics.com/images/zidane.jpg")
img = res[0].plot(show=True)
```
However, there are a couple of issues we need to address:
1. We need to inject the class names and metadata.
2. The bounding boxes seem incorrect, or the model's performance is significantly poor.

![image](https://github.com/user-attachments/assets/8b6b389f-11e7-4a9d-b105-743c16375ae7)

Burhan-Q commented:
Maybe the x-axis coordinates are flipped? The result from `class 18` and `class 119` looks like it _might_ align if you flip the x-coordinates along the vertical axis.

ambitious-octopus committed:
updated mct inference

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
[no ci] updated mct inference

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
[skip ci] updated mct inference

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
ruff skip line

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
Update exporter.py

ambitious-octopus committed:
changed filename

Laughing-q committed:
Update autobackend.py

Laughing-q committed:
Update autobackend.py

Laughing-q committed:
Update exporter.py

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q commented:
@ambitious-octopus ok I've made it work.
- update the name of the model by two quantization methods.
- the results from the models of both methods look correct.
```python
from ultralytics import YOLO
model = YOLO("yolov8n.pt")
model.export(format="mct", device=0)

model = YOLO("yolov8n_gptq_mct_model.onnx")
res = model.predict("ultralytics/assets/zidane.jpg")
img = res[0].plot(show=True)
model = YOLO("yolov8n_ptq_mct_model.onnx")
res = model.predict("ultralytics/assets/zidane.jpg")
img = res[0].plot(show=True)
```
output of  `gptq` model:
![image](https://github.com/user-attachments/assets/04bd8098-d66c-4d34-a9c7-0ef64cb7a3c2)
output of `ptq` model:
![image](https://github.com/user-attachments/assets/2073cd5e-a7ea-4eb4-98b4-382c551dbee3)

but still some work to do:
- We prefer not to return split results [in the head part](https://github.com/ultralytics/ultralytics/blob/4aa9f55b8045069567217a1fbec7d23ddc2e225e/ultralytics/nn/modules/head.py#L120) and with xyxy format and do `xyxy2xywh` and a concatenate operation in [autobackend](https://github.com/ultralytics/ultralytics/blob/4aa9f55b8045069567217a1fbec7d23ddc2e225e/ultralytics/nn/autobackend.py#L469-L471). To maintain the consistency it's better we directly return one tensor and the boxes in it is `xywh` format.
- We need to figure a way(maybe add an argument) to export models with each method(currently we export both two methods at once)

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
mct export refactoring

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus reviewed (commented):
@Laughing-q thank you for the great work! However, I believe it's important not to completely eliminate the Sony post-processing step, as it's necessary for conversion. I've made some modifications to the code and added an `nms` parameter. When `nms=True`, the export will include post-processing; otherwise, it won't. We apply the same approach for exporting to CoreML, so there's no need to modify the `config.yaml` for this parameter. Additionally, I introduced a parameter to choose between GPTQ and PTQ, with `gptq=True` indicating the use of GPTQ.

## Example
```python
model = YOLO("yolov8n.pt")
model.export(format="mct", device=0, nms=False, gptq=True)
model = YOLO("yolov8n_mct_model.onnx")
result = model.predict("https://ultralytics.com/images/zidane.jpg")
result[0].plot(show=True, conf=0.9)
```

## Requirements

To test this example you need to install `model-compression-toolkit `through this branch:
```
pip install git+https://github.com/ambitious-octopus/model_optimization
```

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

ambitious-octopus committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

ambitious-octopus committed:
Merge branch 'main' into quan

Laughing-q committed:
Update head.py

Laughing-q committed:
Update exporter.py

Laughing-q committed:
Update tal.py

Laughing-q committed:
update

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Merge branch 'main' into quan

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
added mct dependency

UltralyticsAssistant committed:
Merge branch 'main' into quan

ambitious-octopus reviewed (commented):
@Idan-BenAmi

# Install
1. clone the repo
```
git clone https://github.com/ultralytics/ultralytics
```
2. Switch to the quan branch
```
git checkout quan
```
3. Install the package as editable
```
pip install -e /path/to/ultralytics/clone
```
4. Install mct-nightly
```
pip install mct-nightly
```

# Usage
```python
from ultralytics import YOLO
model = YOLO("yolov8n.pt") # Get yolov8n
model.export(format="mct", device=0, nms=False, gptq=True) # export with gptq, setting to False will export with ptq
model = YOLO("yolov8n_mct_model.onnx") # Open the exported onnx model
result = model.predict("https://ultralytics.com/images/zidane.jpg") # predict with the exported model
result[0].plot(show=True, conf=0.9) # show result!
```

ambitious-octopus committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

ambitious-octopus committed:
updated core config

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
updated core config

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
added format attr

ambitious-octopus committed:
added yolo11 support

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
added dependencies

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
added docs fixed yolo11 CI

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
fixed mct benchmarking

ambitious-octopus committed:
added benchmark support

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
removed unused code

ambitious-octopus committed:
removed yoloworld support

ambitious-octopus committed:
removed end2end support

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
fixed getattr error

ambitious-octopus committed:
fixed benchmark

ambitious-octopus committed:
Merge branch 'main' into quan

lakshanthad committed:
Add doc to mkdocs

Laughing-q committed:
Merge branch 'main' into quan

Laughing-q committed:
Update exporter.py

ambitious-octopus committed:
better documentation

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into quan

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
updated mct version, added imx500 dependency

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
updated docs

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
updated docs

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q commented:
@ambitious-octopus Nice work buddy! I'll fix the conflict. :)

Laughing-q committed:
update

ambitious-octopus committed:
updated broken links

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
revert back to nighly

ambitious-octopus committed:
revert back to nighly autobackend

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
remove file

Laughing-q committed:
update

ambitious-octopus committed:
updated mct version

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
test 2.1.1

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
fixed onnx inference

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
clean up

Laughing-q committed:
Update exporter.py

Laughing-q committed:
Update exporter.py

Laughing-q committed:
Update exporter.py

Laughing-q committed:
Update exporter.py

Laughing-q committed:
fix

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
revert comment

Laughing-q committed:
Update exporter.py

Laughing-q committed:
update torch.fx

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q commented on ultralytics/nn/modules/head.py:
```diff
@@ -28,6 +28,7 @@ class Detect(nn.Module):
     shape = None
     anchors = torch.empty(0)  # init
     strides = torch.empty(0)  # init
+    format = "_"
```
@ambitious-octopus Hey buddy can we eliminate this line? since we don't have it in Detect along the way and nothing seems wrong to me. :) Also set it equals to `_` looks a bit strange.

ambitious-octopus replied:
Hey @Laughing-q, we will have an attribute error if we remove this. Probably we should set this to `None`. 
Try to remove this line and then:
```
from ultralytics import YOLO
model = YOLO("yolov8n.pt")
model.predict()
```

Laughing-q replied:
@ambitious-octopus ok I see! Thanks! I'll take a look tomorrow. :)

Laughing-q committed:
fix conflicts

Laughing-q committed:
add docstring

Laughing-q committed:
Update torch_utils.py

ambitious-octopus committed:
fixed head format

Laughing-q committed:
Update exporter.py

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
updated imx500 converter version

Laughing-q committed:
update

ambitious-octopus committed:
removed nms=False support

ambitious-octopus committed:
chnaged mct to imx500

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
changed docs from mct to imx500

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus reviewed (commented):
# Changed export name from mct to imx500
```python
from ultralytics import YOLO

model = YOLO("yolov8n.pt")
model.export(format="imx500", data="coco8.yaml", gptq=False)

model = YOLO("yolov8n_imx500_model.onnx")
res = model.predict()
res[0].show()
```

ambitious-octopus committed:
added labels file

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
fixed dependency

ambitious-octopus committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

UltralyticsAssistant committed:
Merge branch 'main' into quan

ambitious-octopus committed:
added tests

ambitious-octopus committed:
added docs updated requirements

ambitious-octopus committed:
fixed tests

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

ambitious-octopus committed:
fixed tests and docs

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-update Ultralytics Docs by https://ultralytics.com/actions

ambitious-octopus committed:
fixed tests

ambitious-octopus committed:
fixed tests

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
Merge branch 'main' into quan

ambitious-octopus committed:
fixed tests

ambitious-octopus reviewed (commented):
@lakshanthad Can you do a final check please? Also you should add:
1. Image of the device inference.
2. Benchmarks.
3. The correct picamera2 branch to use.

lakshanthad committed:
Optimize content

lakshanthad committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Add live inference results

Laughing-q committed:
clean up

Laughing-q committed:
revert the onnx providers

Laughing-q committed:
add comment

Laughing-q committed:
Update default.yaml

Laughing-q committed:
Update imx500.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q committed:
Update mkdocs.yml

Laughing-q committed:
remove blank line

Laughing-q committed:
add comments

Laughing-q commented:
@glenn-jocher This PR this ready from my side! Please take a look as soon as possible! Thanks!

ambitious-octopus committed:
added image

ambitious-octopus reviewed (commented):
@lakshanthad Update cover image!

lakshanthad committed:
Optimize alt text

lakshanthad committed:
Optimize install steps

Chizkiyahu commented on docs/en/integrations/imx500.md:
````diff
@@ -0,0 +1,250 @@
+```bash
+git clone -b next https://github.com/raspberrypi/picamera2
+cd picamera2
+pip install -e .
````
```suggestion
pip install -e .  --break-system-packages
```
without the flag  `--break-system-packages` the command will fail

ambitious-octopus replied:
@Chizkiyahu, thanks for that. FYI @lakshanthad

Chizkiyahu commented on docs/en/integrations/imx500.md:
````diff
@@ -0,0 +1,250 @@
+Step 3: Run YOLOv8 object detection, using the labels.txt file that has been generated during the imx500 export.
+
+```bash
+python imx500_object_detection_demo.py --model <path to network.rpk> --bbox-normalization --ignore-dash-labels --bbox-order xy –labels <path to labels.txt>
````
```suggestion
python imx500_object_detection_demo.py --model <path to network.rpk> --fps 25 --bbox-normalization --ignore-dash-labels --bbox-order xy –labels <path to labels.txt>
```
more then 25 fps on this specific network will degrade the performance

ambitious-octopus committed:
Update docs/en/integrations/imx500.md

Co-authored-by: Chizkiyahu Raful <37312901+Chizkiyahu@users.noreply.github.com>

ambitious-octopus committed:
Update docs/en/integrations/imx500.md

Co-authored-by: Chizkiyahu Raful <37312901+Chizkiyahu@users.noreply.github.com>

glenn-jocher commented:
@ambitious-octopus going through this now!

glenn-jocher committed:
Merge branch 'main' into quan

UltralyticsAssistant committed:
fix mkdocs.yml reference sort order

Signed-off-by: UltralyticsAssistant <web@ultralytics.com>

glenn-jocher committed:
Fix capitalization errors

glenn-jocher committed:
Simplify

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher commented:
Export fails for me in Google Colab notebook, also very verbose with many warnings and errors:

<img width="1680" alt="Screenshot 2024-11-10 at 18 14 36" src="https://github.com/user-attachments/assets/c86b7eb7-fe11-44fd-ba2a-945555cb808d">


Full code to reproduce in https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb

INPUT:
```
!pip install git+https://github.com/ultralytics/ultralytics@quan
!yolo export format=imx500 model=yolov8n.pt
```

OUTPUT:
```
WARNING ⚠️ IMX500 only supports int8 export, setting int8=True.
Ultralytics 8.3.28 🚀 Python-3.10.12 torch-2.5.0+cu121 CPU (Intel Xeon 2.00GHz)
WARNING ⚠️ INT8 export requires a missing 'data' arg for calibration. Using default 'data=coco8.yaml'.
YOLOv8n summary (fused): 168 layers, 3,151,904 parameters, 0 gradients, 8.7 GFLOPs

PyTorch: starting from 'yolov8n.pt' with input shape (1, 3, 640, 640) BCHW and output shape(s) ((1, 8400, 4), (1, 8400, 80)) (6.2 MB)
/usr/local/lib/python3.10/dist-packages/sony_custom_layers/pytorch/object_detection/nms.py:120: FutureWarning: `torch.library.impl_abstract` was renamed to `torch.library.register_fake`. Please use that instead; we will remove `torch.library.impl_abstract` in a future version of PyTorch.
  @torch.library.impl_abstract(MULTICLASS_NMS_TORCH_OP_QUALNAME)
openjdk 11.0.24 2024-07-16
OpenJDK Runtime Environment (build 11.0.24+8-post-Ubuntu-1ubuntu322.04)
OpenJDK 64-Bit Server VM (build 11.0.24+8-post-Ubuntu-1ubuntu322.04, mixed mode, sharing)
IMX500: collecting INT8 calibration images from 'data=coco8.yaml'
Scanning /content/datasets/coco8/labels/val.cache... 4 images, 0 backgrounds, 0 corrupt: 100% 4/4 [00:00<?, ?it/s]
IMX500: WARNING ⚠️ >300 images recommended for INT8 calibration, found 4 images.
representative_data_gen generates a batch size of 1 which can be slow for optimization: consider increasing the batch size
representative_data_gen generates a batch size of 1 which can be slow for optimization: consider increasing the batch size
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
Mixed precision enabled.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
WARNING:Model Compression Toolkit:Pytorch model has a parameter or constant Tensor value. This can cause unexpected behaviour when converting the model.
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
ERROR:Model Compression Toolkit:Found duplicate qco types!
Statistics Collection: 4it [00:03,  1.14it/s]

Running quantization parameters search. This process might take some time, depending on the model size and the selected quantization methods.

Calculating quantization parameters: 100% 194/194 [01:13<00:00,  2.64it/s]
Not enough images in representative dataset to generate 10 data points, only 4 were generated
WARNING:Model Compression Toolkit:Not enough images in representative dataset to generate 10 data points, only 4 were generated
Starting to evaluate metrics
100% 64/64 [02:24<00:00,  2.25s/it]
Welcome to the CBC MILP Solver 
Version: 2.10.3 
Build Date: Dec 15 2019 

command line - /usr/local/lib/python3.10/dist-packages/pulp/solverdir/cbc/linux/64/cbc /tmp/a25f81a388864856b83c5306a9349ea6-pulp.mps -sec 60 -timeMode elapsed -branch -printingOptions all -solution /tmp/a25f81a388864856b83c5306a9349ea6-pulp.sol (default strategy 1)
At line 2 NAME          MODEL
At line 3 ROWS
At line 134 COLUMNS
At line 1159 RHS
At line 1289 BOUNDS
At line 1482 ENDATA
Problem MODEL has 129 rows, 256 columns and 576 elements
Coin0008I MODEL read with 0 errors
seconds was changed from 1e+100 to 60
Option for timeMode changed from cpu to elapsed
Continuous objective value is 0.293378 - 0.00 seconds
Cgl0004I processed model has 65 rows, 192 columns (192 integer (192 of which binary)) and 320 elements
Cbc0038I Initial state - 2 integers unsatisfied sum - 0.829583
Cbc0038I Solution found of 0.294352
Cbc0038I Before mini branch and bound, 190 integers at bound fixed and 0 continuous
Cbc0038I Full problem 65 rows 192 columns, reduced to 0 rows 0 columns
Cbc0038I Mini branch and bound did not improve solution (0.01 seconds)
Cbc0038I Round again with cutoff of 0.294245
Cbc0038I Reduced cost fixing fixed 104 variables on major pass 2
Cbc0038I Pass   1: suminf.    0.05310 (2) obj. 0.294245 iterations 3
Cbc0038I Pass   2: suminf.    0.41479 (2) obj. 0.293521 iterations 3
Cbc0038I Pass   3: suminf.    0.94069 (2) obj. 0.293517 iterations 5
Cbc0038I Pass   4: suminf.    0.18729 (2) obj. 0.294245 iterations 2
Cbc0038I Pass   5: suminf.    0.47035 (2) obj. 0.293678 iterations 3
Cbc0038I Pass   6: suminf.    1.40607 (4) obj. 0.294245 iterations 10
Cbc0038I Pass   7: suminf.    1.16235 (4) obj. 0.294245 iterations 2
Cbc0038I Pass   8: suminf.    0.55931 (2) obj. 0.293717 iterations 3
Cbc0038I Pass   9: suminf.    0.58003 (2) obj. 0.294245 iterations 4
Cbc0038I Pass  10: suminf.    0.72035 (2) obj. 0.293964 iterations 3
Cbc0038I Pass  11: suminf.    1.46018 (4) obj. 0.294245 iterations 3
Cbc0038I Pass  12: suminf.    1.34888 (4) obj. 0.294245 iterations 3
Cbc0038I Pass  13: suminf.    0.11486 (2) obj. 0.294061 iterations 3
Cbc0038I Pass  14: suminf.    0.11486 (2) obj. 0.294061 iterations 0
Cbc0038I Pass  15: suminf.    1.08296 (4) obj. 0.294245 iterations 9
Cbc0038I Pass  16: suminf.    0.95449 (4) obj. 0.294245 iterations 1
Cbc0038I Pass  17: suminf.    0.25722 (2) obj. 0.293966 iterations 4
Cbc0038I Pass  18: suminf.    0.25722 (2) obj. 0.293966 iterations 0
Cbc0038I Pass  19: suminf.    0.92961 (4) obj. 0.294245 iterations 3
Cbc0038I Pass  20: suminf.    1.10783 (4) obj. 0.294245 iterations 1
Cbc0038I Pass  21: suminf.    1.10783 (4) obj. 0.294245 iterations 0
Cbc0038I Pass  22: suminf.    0.20167 (2) obj. 0.294008 iterations 3
Cbc0038I Pass  23: suminf.    0.20167 (2) obj. 0.294008 iterations 0
Cbc0038I Pass  24: suminf.    1.07616 (4) obj. 0.294245 iterations 6
Cbc0038I Pass  25: suminf.    0.69247 (4) obj. 0.294245 iterations 4
Cbc0038I Pass  26: suminf.    0.70167 (2) obj. 0.293675 iterations 3
Cbc0038I Pass  27: suminf.    0.70167 (2) obj. 0.293675 iterations 0
Cbc0038I Pass  28: suminf.    0.47603 (2) obj. 0.294245 iterations 3
Cbc0038I Pass  29: suminf.    0.64917 (2) obj. 0.293899 iterations 2
Cbc0038I Pass  30: suminf.    1.72559 (4) obj. 0.294245 iterations 5
Cbc0038I No solution found this major pass
Cbc0038I Before mini branch and bound, 168 integers at bound fixed and 0 continuous
Cbc0038I Full problem 65 rows 192 columns, reduced to 1 rows 12 columns
Cbc0038I Mini branch and bound improved solution from 0.294352 to 0.29371 (0.01 seconds)
Cbc0038I Round again with cutoff of 0.293636
Cbc0038I Reduced cost fixing fixed 114 variables on major pass 3
Cbc0038I Pass  30: suminf.    0.61004 (2) obj. 0.293636 iterations 1
Cbc0038I Pass  31: suminf.    0.82958 (2) obj. 0.293378 iterations 9
Cbc0038I Pass  32: suminf.    1.18773 (4) obj. 0.293636 iterations 3
Cbc0038I Pass  33: suminf.    0.63931 (4) obj. 0.293636 iterations 1
Cbc0038I Pass  34: suminf.    0.74278 (2) obj. 0.293454 iterations 1
Cbc0038I Pass  35: suminf.    0.58776 (2) obj. 0.293636 iterations 1
Cbc0038I Pass  36: suminf.    0.09841 (2) obj. 0.293636 iterations 2
Cbc0038I Solution found of 0.293631
Cbc0038I Before mini branch and bound, 182 integers at bound fixed and 0 continuous
Cbc0038I Full problem 65 rows 192 columns, reduced to 1 rows 3 columns
Cbc0038I Mini branch and bound did not improve solution (0.02 seconds)
Cbc0038I Round again with cutoff of 0.293548
Cbc0038I Reduced cost fixing fixed 117 variables on major pass 4
Cbc0038I Pass  37: suminf.    0.68502 (2) obj. 0.293548 iterations 0
Cbc0038I Pass  38: suminf.    0.82958 (2) obj. 0.293378 iterations 9
Cbc0038I Pass  39: suminf.    0.45507 (4) obj. 0.293548 iterations 3
Cbc0038I Pass  40: suminf.    0.45507 (4) obj. 0.293548 iterations 0
Cbc0038I Pass  41: suminf.    0.32909 (2) obj. 0.293548 iterations 1
Cbc0038I Pass  42: suminf.    0.25722 (2) obj. 0.293464 iterations 1
Cbc0038I Pass  43: suminf.    0.25722 (2) obj. 0.293464 iterations 0
Cbc0038I Pass  44: suminf.    0.25722 (2) obj. 0.293464 iterations 0
Cbc0038I Pass  45: suminf.    0.25722 (2) obj. 0.293464 iterations 1
Cbc0038I Pass  46: suminf.    1.39584 (4) obj. 0.293548 iterations 2
Cbc0038I Pass  47: suminf.    1.38626 (4) obj. 0.293548 iterations 1
Cbc0038I Pass  48: suminf.    0.49625 (2) obj. 0.293508 iterations 1
Cbc0038I Pass  49: suminf.    0.46192 (2) obj. 0.293548 iterations 1
Cbc0038I Pass  50: suminf.    0.73885 (4) obj. 0.293548 iterations 5
Cbc0038I Pass  51: suminf.    0.19313 (2) obj. 0.293535 iterations 1
Cbc0038I Pass  52: suminf.    0.89909 (2) obj. 0.293548 iterations 2
Cbc0038I Pass  53: suminf.    0.79542 (2) obj. 0.293426 iterations 1
Cbc0038I Pass  54: suminf.    0.79542 (2) obj. 0.293426 iterations 1
Cbc0038I Pass  55: suminf.    0.79542 (2) obj. 0.293426 iterations 0
Cbc0038I Pass  56: suminf.    0.68431 (2) obj. 0.293529 iterations 3
Cbc0038I Pass  57: suminf.    0.68431 (2) obj. 0.293529 iterations 0
Cbc0038I Pass  58: suminf.    0.70011 (2) obj. 0.293548 iterations 1
Cbc0038I Pass  59: suminf.    0.76156 (2) obj. 0.293548 iterations 4
Cbc0038I Pass  60: suminf.    0.73986 (2) obj. 0.293522 iterations 1
Cbc0038I Pass  61: suminf.    0.76156 (2) obj. 0.293548 iterations 1
Cbc0038I Pass  62: suminf.    1.76940 (4) obj. 0.293548 iterations 3
Cbc0038I Pass  63: suminf.    1.76940 (4) obj. 0.293548 iterations 0
Cbc0038I Pass  64: suminf.    0.91976 (2) obj. 0.293548 iterations 1
Cbc0038I Pass  65: suminf.    0.88222 (2) obj. 0.293504 iterations 1
Cbc0038I Pass  66: suminf.    0.79542 (2) obj. 0.293426 iterations 1
Cbc0038I No solution found this major pass
Cbc0038I Before mini branch and bound, 174 integers at bound fixed and 0 continuous
Cbc0038I Full problem 65 rows 192 columns, reduced to 1 rows 8 columns
Cbc0038I Mini branch and bound improved solution from 0.293631 to 0.293606 (0.02 seconds)
Cbc0038I Round again with cutoff of 0.293497
Cbc0038I Reduced cost fixing fixed 117 variables on major pass 5
Cbc0038I Pass  66: suminf.    0.72839 (2) obj. 0.293497 iterations 0
Cbc0038I Pass  67: suminf.    0.82958 (2) obj. 0.293378 iterations 9
Cbc0038I Pass  68: suminf.    1.08355 (4) obj. 0.293497 iterations 3
Cbc0038I Pass  69: suminf.    1.08355 (4) obj. 0.293497 iterations 0
Cbc0038I Pass  70: suminf.    0.15599 (2) obj. 0.293497 iterations 1
Cbc0038I Pass  71: suminf.    0.11486 (2) obj. 0.293449 iterations 1
Cbc0038I Pass  72: suminf.    0.52908 (4) obj. 0.293497 iterations 2
Cbc0038I Pass  73: suminf.    0.11486 (2) obj. 0.293484 iterations 1
Cbc0038I Pass  74: suminf.    0.12592 (2) obj. 0.293497 iterations 1
Cbc0038I Pass  75: suminf.    0.12592 (2) obj. 0.293497 iterations 0
Cbc0038I Pass  76: suminf.    0.71275 (4) obj. 0.293497 iterations 2
Cbc0038I Pass  77: suminf.    0.52530 (4) obj. 0.293497 iterations 3
Cbc0038I Pass  78: suminf.    0.20167 (2) obj. 0.293485 iterations 1
Cbc0038I Pass  79: suminf.    0.21192 (2) obj. 0.293497 iterations 1
Cbc0038I Pass  80: suminf.    1.55519 (4) obj. 0.293497 iterations 3
Cbc0038I Pass  81: suminf.    0.62011 (2) obj. 0.293497 iterations 1
Cbc0038I Pass  82: suminf.    0.65597 (2) obj. 0.293455 iterations 1
Cbc0038I Pass  83: suminf.    0.34403 (2) obj. 0.293465 iterations 5
Cbc0038I Pass  84: suminf.    0.34403 (2) obj. 0.293465 iterations 0
Cbc0038I Pass  85: suminf.    0.37173 (2) obj. 0.293497 iterations 1
Cbc0038I Pass  86: suminf.    1.18796 (4) obj. 0.293497 iterations 4
Cbc0038I Pass  87: suminf.    0.25722 (2) obj. 0.293464 iterations 1
Cbc0038I Pass  88: suminf.    0.28572 (2) obj. 0.293497 iterations 1
Cbc0038I Pass  89: suminf.    0.25722 (2) obj. 0.293464 iterations 3
Cbc0038I Pass  90: suminf.    0.28572 (2) obj. 0.293497 iterations 1
Cbc0038I Pass  91: suminf.    0.17020 (2) obj. 0.293491 iterations 4
Cbc0038I Pass  92: suminf.    0.17020 (2) obj. 0.293491 iterations 0
Cbc0038I Pass  93: suminf.    0.17569 (2) obj. 0.293497 iterations 1
Cbc0038I Pass  94: suminf.    0.71961 (4) obj. 0.293497 iterations 3
Cbc0038I Pass  95: suminf.    0.46378 (4) obj. 0.293497 iterations 1
Cbc0038I No solution found this major pass
Cbc0038I Before mini branch and bound, 172 integers at bound fixed and 0 continuous
Cbc0038I Full problem 65 rows 192 columns, reduced to 1 rows 9 columns
Cbc0038I Mini branch and bound did not improve solution (0.03 seconds)
Cbc0038I After 0.03 seconds - Feasibility pump exiting with objective of 0.293606 - took 0.02 seconds
Cbc0012I Integer solution of 0.2936064 found by feasibility pump after 0 iterations and 0 nodes (0.03 seconds)
Cbc0012I Integer solution of 0.29352974 found by DiveCoefficient after 0 iterations and 0 nodes (0.03 seconds)
Cbc0031I 3 added rows had average density of 12.333333
Cbc0013I At root node, 11 cuts changed objective from 0.29337833 to 0.29352342 in 2 passes
Cbc0014I Cut generator 0 (Probing) - 4 row cuts average 2.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is 1
Cbc0014I Cut generator 1 (Gomory) - 0 row cuts average 0.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is -100
Cbc0014I Cut generator 2 (Knapsack) - 3 row cuts average 4.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is 1
Cbc0014I Cut generator 3 (Clique) - 0 row cuts average 0.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is -100
Cbc0014I Cut generator 4 (MixedIntegerRounding2) - 0 row cuts average 0.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is -100
Cbc0014I Cut generator 5 (FlowCover) - 0 row cuts average 0.0 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is -100
Cbc0014I Cut generator 6 (TwoMirCuts) - 12 row cuts average 15.5 elements, 0 column cuts (0 active)  in 0.000 seconds - new frequency is -100
Cbc0001I Search completed - best objective 0.2935297363844492, took 6 iterations and 0 nodes (0.03 seconds)
Cbc0035I Maximum depth 0, 119 variables fixed on reduced cost
Cuts at root node changed objective from 0.293378 to 0.293523
Probing was tried 2 times and created 4 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)
Gomory was tried 2 times and created 0 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)
Knapsack was tried 2 times and created 3 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)
Clique was tried 2 times and created 0 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)
MixedIntegerRounding2 was tried 2 times and created 0 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)
FlowCover was tried 2 times and created 0 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)
TwoMirCuts was tried 2 times and created 12 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)
ZeroHalf was tried 1 times and created 0 cuts of which 0 were active after adding rounds of cuts (0.000 seconds)

Result - Optimal solution found

Objective value:                0.29352974
Enumerated nodes:               0
Total iterations:               6
Time (CPU seconds):             0.03
Time (Wallclock seconds):       0.04

Option for printingOptions changed from normal to all
Total time (CPU seconds):       0.03   (Wallclock seconds):       0.04

Optimal
Final weights bit-width configuration: [8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 4, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 8, 4, 8, 8, 8, 8, 8, 8, 4, 8, 4, 4, 4, 8, 8, 4, 4, 4, 4, 4, 4, 4, 4, 8, 8, 8]

Please run your accuracy evaluation on the exported quantized model to verify it's accuracy.
Checkout the FAQ and Troubleshooting pages for resolving common issues and improving the quantized model accuracy:
FAQ: https://github.com/sony/model_optimization/tree/main/FAQ.md
Quantization Troubleshooting: https://github.com/sony/model_optimization/tree/main/quantization_troubleshooting.md
Exporting onnx model with MCTQ quantizers: yolov8n_imx500_model.onnx
2024-11-10 17:13:45,333 INFO : Running version 1.10.0 [/usr/local/lib/python3.10/dist-packages/uni/common/logger.py:148]
2024-11-10 17:13:45,334 INFO : Converting yolov8n_imx500_model.onnx [/usr/local/lib/python3.10/dist-packages/uni/common/logger.py:148]
2024-11-10 17:14:08,596 INFO : Wrote outputs to /tmp/tmpbq2bivoc/yolov8n_imx500_model.uni-pytorch.um.pb [/usr/local/lib/python3.10/dist-packages/uni/common/logger.py:148]
2024-11-10 17:14:08,632 INFO : Converted successfully [/usr/local/lib/python3.10/dist-packages/uni/common/logger.py:148]
2024-11-10 17:14:08 WARNING : Java 17 not found!
2024-11-10 17:14:08 WARNING : Please make sure you have Java 17 installed and configured correctly.
2024-11-10 17:14:08 WARNING : You can set JAVA_HOME environment variable to point to the directory where Java 17 is installed.
2024-11-10 17:14:08 WARNING : Example: 
2024-11-10 17:14:08 WARNING :   export JAVA_HOME=/path/to/java/17
2024-11-10 17:14:08 WARNING :   export PATH=$JAVA_HOME/bin:$PATH
2024-11-10 17:14:08 WARNING : You can suppress this warning by using the '--no-verify-java' flag.
Error: LinkageError occurred while loading main class sdsp.app.AppKt
	java.lang.UnsupportedClassVersionError: sdsp/app/AppKt has been compiled by a more recent version of the Java Runtime (class file version 61.0), this version of the Java Runtime only recognizes class file versions up to 55.0
2024-11-10 17:14:08,791 ERROR : sdspconv exited with error code: 1, uni files can be found in /tmp/tmpbq2bivoc [/usr/local/lib/python3.10/dist-packages/imx500_converter/converter/converter.py:53]
IMX500: export failure ❌ 267.1s: Command '['imxconv-pt', '-i', 'yolov8n_imx500_model.onnx', '-o', 'yolov8n_imx500_model', '--no-input-persistency', '--overwrite-output']' returned non-zero exit status 1.
Traceback (most recent call last):
  File "/usr/local/bin/yolo", line 8, in <module>
    sys.exit(entrypoint())
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/cfg/__init__.py", line 966, in entrypoint
    getattr(model, mode)(**overrides)  # default args from model
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/model.py", line 734, in export
    return Exporter(overrides=args, _callbacks=self.callbacks)(model=self.model)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/exporter.py", line 380, in __call__
    f[13], _ = self.export_imx500()
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/exporter.py", line 147, in outer_func
    raise e
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/exporter.py", line 142, in outer_func
    f, model = inner_func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ultralytics/engine/exporter.py", line 1217, in export_imx500
    subprocess.run(
  File "/usr/lib/python3.10/subprocess.py", line 526, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['imxconv-pt', '-i', 'yolov8n_imx500_model.onnx', '-o', 'yolov8n_imx500_model', '--no-input-persistency', '--overwrite-output']' returned non-zero exit status 1.
Sentry is attempting to send 2 pending events
Waiting up to 2 seconds
Press Ctrl-C to quit
```

Also why do we have two calls to `check_requirements()` instead of just one @ambitious-octopus?

ambitious-octopus committed:
chnaged naming from imx500 to imx

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
moved onnx file inside imx folder

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
fixed tables

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

lakshanthad committed:
Add Benchmarks table

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

ambitious-octopus committed:
fixed benchmarks

ambitious-octopus committed:
added newline

ambitious-octopus committed:
standardize exporter

ambitious-octopus committed:
fixed docs

ambitious-octopus committed:
fixed path

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

glenn-jocher committed:
Update benchmarks.py

glenn-jocher committed:
Update head.py

glenn-jocher committed:
Update head.py

glenn-jocher commented on ultralytics/engine/exporter.py:
```diff
@@ -1068,6 +1099,135 @@ def export_tfjs(self, prefix=colorstr("TensorFlow.js:")):
+        check_requirements(("model-compression-toolkit==2.1.1", "sony-custom-layers==0.2.0", "tensorflow==2.12.0"))
+        check_requirements("imx500-converter[pt]==3.14.3")  # Separate requirements for imx500-converter
+
+        import subprocess
```
@ambitious-octopus this import is redundant

<img width="485" alt="Screenshot 2024-11-11 at 00 08 16" src="https://github.com/user-attachments/assets/cf563262-b065-43b1-a935-7e7a883fa688">

glenn-jocher commented on ultralytics/engine/exporter.py:
```diff
@@ -1068,6 +1099,135 @@ def export_tfjs(self, prefix=colorstr("TensorFlow.js:")):
+            check=True,
+        )
+
+        # Needed for imx models.
+        with open(f / "labels.txt", "w") as file:
+            file.writelines([f"{name}\n" for _, name in self.model.names.items()])
```
@ambitious-octopus this is a crime. The ONNX models have the Ultralytics metadata embedded inside them. The IMX models are incapable of using this?

ambitious-octopus replied:
@glenn-jocher The official imx500 [model deployment script] (https://github.com/raspberrypi/picamera2/blob/main/examples/imx500/imx500_object_detection_demo.py) reads labels from a `.txt` file. Also, the converter (`imx500-converter`) does not take this metadata during conversion. For this reason, we added a `.txt` file with labels.

glenn-jocher commented on ultralytics/nn/autobackend.py:
```diff
@@ -199,7 +200,22 @@ def __init__(
+                check_requirements(
+                    ["model-compression-toolkit==2.1.1", "sony-custom-layers[torch]==0.2.0", "onnxruntime-extensions"]
+                )
+                w = [i for i in Path(w).iterdir() if i.name.endswith("onnx")][0]
```
No, this is not the right way to find a file in a directory. See OpenVINO example that finds the *.xml file in the directory with glob.

glenn-jocher commented on ultralytics/nn/autobackend.py:
```diff
@@ -199,7 +200,22 @@ def __init__(
                 cuda = False
             LOGGER.info(f"Preferring ONNX Runtime {providers[0]}")
-            session = onnxruntime.InferenceSession(w, providers=providers)
+            if imx:
```
@ambitious-octopus `if else` statements should always have the most commonly expected first result first. In this case that's `if onnx`, NOT `if imx`

glenn-jocher commented on ultralytics/engine/exporter.py:
```diff
@@ -273,6 +290,18 @@ def __call__(self, model=None) -> str:
+            if isinstance(m, Detect) and imx:
+                from ultralytics.utils.tal import make_anchors
+
+                m.anchors, m.strides = (
+                    x.transpose(0, 1)
+                    for x in make_anchors(
+                        torch.cat([s / m.stride.unsqueeze(-1) for s in self.imgsz], dim=1), m.stride, 0.5
+                    )
+                )
```
@ambitious-octopus @Laughing-q why are anchors being manually computed here when they are computed in the forward pass anyways?

Laughing-q replied:
@glenn-jocher Sony export is based on torch.fx. They use torch.fx to export the static graph of the model and do quantization based on the graph. And unfortunately we need this to make torch.fx work.
Rules for torch.fx to export the static graph is that: 
- It takes the input as `symbol` rather than a real data, so we're not able to access any of the attributes from input. And our anchors generation in head.py is using its dtype and device so it does not work with it.
  - To resolve and I tried not using dtype and device in the function but for that we need to explicitly put the generated anchors to correct device which we can't get the correct device in this Detect.forward method because we can not access attributes like `device` of the input.
- We also can not use dynamic control flow(where the condition may depend on the value of input, again it's a symbol, not real number so can not have its true value). See https://pytorch.org/docs/stable/fx.html#dynamic-control-flow.
  - which mean https://github.com/ultralytics/ultralytics/pull/14878/commits/50aba1e908dbcd2e5d77ccd45f45e70b71a7c97a has already broken the workflow, since we need the format at the front so it doesn't access `x[0].shape` when export imx format.

Laughing-q replied:
@glenn-jocher I've reverted the updates that breaks the workflow in https://github.com/ultralytics/ultralytics/pull/14878/commits/4c36795116779ef8407f6a6188d9111bbc58dcb2 for now

glenn-jocher commented on ultralytics/nn/modules/block.py:
```diff
@@ -238,6 +238,13 @@ def forward(self, x):
         y.extend(m(y[-1]) for m in self.m)
         return self.cv2(torch.cat(y, 1))
 
+    def forward_fx(self, x):
+        """Forward pass creating list explicitly instead of using list()."""
+        y1 = self.cv1(x).chunk(2, 1)
+        y = [y1[0], y1[1]]
+        y.extend(m(y[-1]) for m in self.m)
+        return self.cv2(torch.cat(y, 1))
```
@ambitious-octopus @Laughing-q forward_split() below is used for export for all formats including ONNX. Don't you want to modify this method instead of forward()?

Laughing-q replied:
@glenn-jocher make sense! did the update in https://github.com/ultralytics/ultralytics/pull/14878/commits/5fbcdb1bb0f9d689b8cc3cce102603b625f1f806

glenn-jocher committed:
Update exporter.py

@ambitious-octopus this is just a normal check and belongs in the checks section, not specially placed way ahead of everything else.

glenn-jocher committed:
Update sony-imx500.md

glenn-jocher committed:
Fix equals

glenn-jocher committed:
Merge branch 'main' into quan

glenn-jocher commented:
All, I've detached the docs page from the code in https://github.com/ultralytics/ultralytics/pull/17452 and merged this now. This new docs page is live at https://docs.ultralytics.com/integrations/sony-imx500. Let's try to sort out the remaining code issues Monday.

glenn-jocher committed:
Update sony-imx500.md

Laughing-q committed:
revert format condition

Laughing-q committed:
Update exporter.py

Laughing-q committed:
remove forward_fx

Laughing-q committed:
Update exporter.py

Laughing-q committed:
create FXModel and remove torchfx

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q commented:
@glenn-jocher ok I created a new class `FXModel` specifically for `export_imx` to get rid of `*args` and `**kwargts` and removed `torchfx`. And it works properly with torch.fx. https://github.com/ultralytics/ultralytics/pull/14878/commits/a917fdebd07857adb5a8ba33b11ce52f18c751f8

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

Laughing-q committed:
clean up

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

UltralyticsAssistant committed:
Auto-update Ultralytics Docs Reference by https://ultralytics.com/actions

Laughing-q committed:
update

Laughing-q committed:
clean up

ambitious-octopus committed:
revert int8 check

ambitious-octopus committed:
added support for colab

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

Laughing-q commented:
@glenn-jocher @ambitious-octopus Guys after looking into the if-else statement, actually I think it has to be the way it looks like now.
```python
if self.format != "imx" and (self.dynamic or self.shape != shape):
```
since we need to stop torch.fx from accessing the `shape` in the statement, and it has to be `and` in the middle. 
If we get self.export involved then the statement need to be:
```python
if (self.export and self.format != "imx") or self.dynamic or self.shape != shape:
```
And it need `or` in the middle or the anchors generation won't work when `self.export=False`, but torch.fx would try to access the latter statements even if `self.export and self.format != "imx"` is `False`, hence it accesses the `shape`...

As we need to stop the torch.fx trace from accessing `shape` we need/have to update this if-else statement(because it's the first one) instead of trying to modify other ones...

@glenn-jocher or is there a way to remove `self.shape == shape`? might make it more complicated as we're having it all the time.

RizwanMunawar committed:
Add sony-imx500.md to mkdocs.yml (#17463)

Co-authored-by: Mohammed Yasin <32206511+Y-T-G@users.noreply.github.com>
Co-authored-by: Glenn Jocher <glenn.jocher@ultralytics.com>

glenn-jocher committed:
Merge branch 'main' into quan

RizwanMunawar committed:
Update index.md

Add sony-imx500 in integrations/index.md

UltralyticsAssistant committed:
Auto-format by https://ultralytics.com/actions

RizwanMunawar committed:
Update index.md

glenn-jocher commented:
@Laughing-q got it! I'll merge this as-is and we can revisit this later as I think there must be a better way to handle this, but also don't the answer right now.

glenn-jocher committed:
Update __init__.py

Laughing-q commented:
@glenn-jocher Agreed!

glenn-jocher commented:
@Laughing-q @ambitious-octopus @lakshanthad @RizwanMunawar PR merged!! Great work everyone :)
